src/bin/nydusd/fs_cache.rs

```diff
@@ -604,38 +604,44 @@ impl FsCacheHandler {
                 }
             }
             Some((FsCacheObject::Bootstrap(bs), u)) => {
-                // TODO: should we feed the bootstrap at together to improve performance?
                 fd = u;
-                let base = unsafe {
-                    libc::mmap(
-                        std::ptr::null_mut(),
-                        msg.len as usize,
-                        libc::PROT_READ,
-                        libc::MAP_SHARED,
-                        bs.bootstrap_file.as_raw_fd(),
-                        msg.off as libc::off_t,
-                    )
-                };
-                if base == libc::MAP_FAILED {
-                    warn!(
-                        "fscache: failed to mmap bootstrap file, {}",
-                        std::io::Error::last_os_error()
-                    );
-                } else {
-                    let ret = unsafe {
-                        libc::pwrite(
-                            bs.cache_file.as_raw_fd(),
-                            base,
-                            msg.len as usize,
-                            msg.off as libc::off_t,
-                        )
-                    };
-                    let _ = unsafe { libc::munmap(base, msg.len as usize) };
-                    if ret < 0 {
-                        warn!(
-                            "fscache: failed to write bootstrap blob data to cached file, {}",
-                            std::io::Error::last_os_error()
-                        );
+                match bs.bootstrap_file.metadata() {
+                    Err(e) => {
+                        warn!("fscache: get bootstrap file metadata err {}", e);
+                    }
+                    Ok(md) => {
+                        let base = unsafe {
+                            libc::mmap(
+                                std::ptr::null_mut(),
+                                md.len() as usize,
+                                libc::PROT_READ,
+                                libc::MAP_SHARED,
+                                bs.bootstrap_file.as_raw_fd(),
+                                0_i64,
+                            )
+                        };
+                        if base == libc::MAP_FAILED {
+                            warn!(
+                                "fscache: failed to mmap bootstrap file, {}",
+                                std::io::Error::last_os_error()
+                            );
+                        } else {
+                            let ret = unsafe {
+                                libc::pwrite(
+                                    bs.cache_file.as_raw_fd(),
+                                    base,
+                                    md.len() as usize,
+                                    0_i64,
```

Comment:
`pwrite` can't guarantee that all `md.len()` bytes of data are written to the file, so a more reliable way is to check if `ret` is equal to `md.len()` and give warn then if fscache on-demand load can't handle such exception.

Reply:
Maybe, we can just fulfill bootstrap in an async context thus not affecting this EROFS IO.

Reply:
> Maybe, we can just fulfill bootstrap in an async context thus not affecting this EROFS IO.
Yes , I think this should be better , we need to make sure all fscache cache files (include bootstrap)fully filled before nydusd exit. So we can launch this async context when handling open requests for bootstrap , does this make sense ?

Reply:
Sounds good to me.